src/stream/quic_lb/ngx_stream_quic_comm.c

```diff
@@ -195,3 +195,180 @@ ngx_quic_aes_128_ecb_encrypt(u_char *plaintext, ngx_int_t plaintext_len,
     EVP_CIPHER_CTX_free(ctx);
     return NGX_ERROR;
 }
+
+
+ngx_int_t
+ngx_quic_aes_128_ecb_decrypt(u_char *ciphertext, ngx_int_t ciphertext_len,
+    u_char *key, u_char *plaintext)
+{
+    EVP_CIPHER_CTX   *ctx;
+    int               len;
+    ngx_int_t         plaintext_len;
+
+    ctx = EVP_CIPHER_CTX_new();
+    if (ctx == NULL) {
+        goto failed;
+    }
+
+    if (EVP_DecryptInit_ex(ctx, EVP_aes_128_ecb(), NULL, key, NULL) <= 0 ) {
+        goto failed;
+    }
+
+    if (EVP_CIPHER_CTX_set_padding(ctx, 0) <= 0) {
+        goto failed;
+    }
+
+    if (EVP_DecryptUpdate(ctx, plaintext, &len, ciphertext, ciphertext_len) <= 0) {
+        goto failed;
+    }
+    plaintext_len = len;
+
+    if (EVP_DecryptFinal_ex(ctx, plaintext + len, &len) <= 0) {
+        goto failed;
+    }
+    plaintext_len += len;
+
+    EVP_CIPHER_CTX_free(ctx);
+    return plaintext_len;
+failed:
+    EVP_CIPHER_CTX_free(ctx);
+    return NGX_ERROR;
+}
+
+ngx_int_t
+ngx_quic_expand_left(u_char *result, u_char *s1, ngx_int_t s1_bits,
+    u_char *s2, ngx_int_t s2_bits)
+{
+    ngx_int_t i, j, offset = 0;
+    ngx_int_t s1_byte, s1_bitofbyte, s2_byte, s2_bitofbyte;
+
+
+    if (s1_bits + s2_bits > 128) {
+        return NGX_ERROR;
+    }
+
+    ngx_memzero(result, NGX_QUIC_LB_STREAM_CIPHER_PADDED_DATA_LEN);
+
+    s1_byte = s1_bits / 8;
+    s1_bitofbyte = s1_bits % 8;
+    s2_byte = s2_bits / 8;
+    s2_bitofbyte = s2_bits % 8;
+
+    for (i = 0; i < s1_byte; i++) {
+        result[i] = s1[i];
+    }
+
+    for (j = 0; j < s1_bitofbyte; j++) {
+        result[i] |= (s1[i]) & (1 << (7 - j));
+    }
+
+    if (s2_bitofbyte != 0) {
+        offset = 1;
+    }
+
+    for (i = NGX_QUIC_LB_STREAM_CIPHER_PADDED_DATA_LEN - 1;
+         i > NGX_QUIC_LB_STREAM_CIPHER_PADDED_DATA_LEN - 1 - s2_byte; i--) {
+        result[i] = s2[s2_byte + i - NGX_QUIC_LB_STREAM_CIPHER_PADDED_DATA_LEN + offset];
+    }
+
+    for (j = 0; j < s2_bitofbyte; j++) {
+        result[i] |= (s2[0]) & (1 << j);
+    }
```

Comment:
I suggest that avoid doing assignment through loop, it's hard to understand and ineffective. the function `expand_lent` could be implemented like this:
```
memzero(result)

memcpy(result, s1, s1_byte)

if (s1_bitofbyte > 0 ) {
    result[s1_byte-1] |= (s1[s1_byte-1]) & 0xf0;
}

if (s2_bitofbyte != 0) {
        offset = 1;
    }

memcpy(result+NGX_QUIC_LB_STREAM_CIPHER_PADDED_DATA_LEN-s2_byte-offset, s2, s2_byte+offset)

if (s2_bitofbyte != 0 ) {
    result[NGX_QUIC_LB_STREAM_CIPHER_PADDED_DATA_LEN-s2_byte-offset] &= 0x0f;
}
```